### ISSUE TITLE
Refactor Input Manager to work with EventReader/Writer

### ISSUE THREAD

**Issue by RiscadoA** [GameDevTecnico/cubos#208]
At the moment the Input Manager emits information only through callbacks, which are fundamentally incompatible with ECS philosophy.
The solution is to instead emit events through #202. Basic window events should also be written to these objects, and read from them.

**Comment by RiscadoA** [comment]
Depends on #285.

**Comment by RiscadoA** [comment]
# New Design

A new resource `Input` is added, which contains all of the current input state (basically renamed `InputManager`).
Through `Input`, the user should be able to:
- get the state of actions through their names (pressed, released, hold).
- get the value of axes through their names (e.g., "mouse.x", "mouse.y").
- remove, replace or add different types of bindings from/to actions.
- remove, replace or add different types of sources from/to axes.

## Actions

Actions correspond to possible user actions like 'jump', 'shoot'.

## Axes

Axes correspond to user input which has a real value: 'move forward' could range from -1 to 1 for example.
Each axis has four properties:
- minimum value;
- maximum value;
- default: value desired when no source is active;
- gravity: how fast the value falls towards the default when no source is active (can be infinite for no delay);

Extra properties could also be added to handle other cases such as:
- what to do if more than one source is active (average, choose first);
- whether to move smoothly to the value given by the sources;

## Bindings

Bindings are the actual inputs that trigger actions.

TODO

## Sources

Sources are the actual inputs that supply values to axes.
All source types derive from a base class `Source`.

Source provides two virtual methods: 

TODO

**Comment by RiscadoA** [comment]
Yes, the idea is that both systems are available. We still have the `WindowEvent`s available, and with this input redesign I was imagining something like `InputEvent` which would be triggered by actions being pressed/released. We could also add events for the axes.

And yeah, the `gravity` property can be implemented manually by the user, I just added it because its something you have at least in [Unity](https://docs.unity3d.com/Manual/class-InputManager.html). (its also super easy to implement)

**Comment by RiscadoA** [comment]
No longer relevant, the Input plugin is being completely redesigned.
